src/ethereum/shanghai/fork.py

```diff
@@ -708,14 +715,18 @@ def calculate_intrinsic_cost(tx: Transaction) -> Uint:
     -------
     verified : `ethereum.base_types.Uint`
         The intrinsic cost of the transaction.
+    verified : `ethereum.base_types.Uint`
+        The calldata tokens used.
     """
     data_cost = 0
-
+    zero_bytes = 0
     for byte in tx.data:
         if byte == 0:
-            data_cost += TX_DATA_COST_PER_ZERO
-        else:
-            data_cost += TX_DATA_COST_PER_NON_ZERO
+            zero_bytes += 1
+            
+    tokens_in_calldata = zero_bytes + (len(tx.data) - zero_bytes) * 4
+    
+    data_cost = tokens_in_calldata * TOTAL_COST_FLOOR_PER_TOKEN
```

Comment:
It doesn't make sense to do https://github.com/ethereum/execution-specs/blob/f34da0bf741a43345229cd3d0567f95b419dfe9a/src/ethereum/shanghai/fork.py#L619

If the `data_cost =  tokens_in_calldata * TOTAL_COST_FLOOR_PER_TOKEN`

the gas_used is always gonna be greater than `floor`

explaination:

`floor = tokens_in_calldata * TOTAL_COST_FLOOR_PER_TOKEN + TX_BASE_COST ` 
i.e
`floor = data_cost + TX_BASE_COST ` 

And  `intrinsic_cost = TX_BASE_COST + data_cost + something`
Or `intrinsic_cost = floor + something`
Or `intrinsic_cost >= floor `

and `gas_used = intrinsic_cost + something`
Or  `gas_used >= floor` always

Reply:
Accidentally closed. will reopen and fix that